### ISSUE TITLE
error in AMQP output plugin when connection is reset

### ISSUE THREAD

**Issue by zstyblik** [mozilla-services/heka#1756]
Hello,

we've hit rather strange error with AMQP output plugin. Log output follows:

```
Sep 07 06:19:47 some-server hekad[9695]: 2015/09/07 06:19:47 Input started: JsonSocketInput
Oct 05 16:55:17 some-server hekad[9695]: 2015/10/05 16:55:17 Diagnostics: 32 packs have been idle more than 120 seconds.
Oct 05 16:55:17 some-server hekad[9695]: 2015/10/05 16:55:17 Diagnostics: (input) Plugin names and quantities found on idle packs:
Oct 05 16:55:17 some-server hekad[9695]: 2015/10/05 16:55:17 Diagnostics:         AMQPOutput: 32
Oct 05 16:55:17 some-server hekad[9695]: 2015/10/05 16:55:17
[...]
Oct 05 17:06:47 some-server hekad[9695]: 2015/10/05 17:06:47 Diagnostics: 32 packs have been idle more than 120 seconds.
Oct 05 17:06:47 some-server hekad[9695]: 2015/10/05 17:06:47 Diagnostics: (input) Plugin names and quantities found on idle packs:
Oct 05 17:06:47 some-server hekad[9695]: 2015/10/05 17:06:47 Diagnostics:         AMQPOutput: 32
Oct 05 17:06:47 some-server hekad[9695]: 2015/10/05 17:06:47
Oct 06 07:06:41 some-server hekad[9695]: 2015/10/06 07:06:41 Plugin 'AMQPOutput' error: write tcp AA.BB.CC.DD:5672: connection reset by peer
Oct 06 07:06:41 some-server hekad[9695]: 2015/10/06 07:06:41 Plugin 'AMQPOutput': stopped
Oct 06 07:06:42 some-server hekad[9695]: 2015/10/06 07:06:42 Plugin 'AMQPOutput': now restarting
Oct 06 07:06:42 some-server hekad[9695]: 2015/10/06 07:06:42 Plugin 'AMQPOutput' error: AMQP scheme must be either 'amqp://' or 'amqps://'
Oct 06 07:06:42 some-server hekad[9695]: 2015/10/06 07:06:42 Plugin 'AMQPOutput': now restarting
[...]
```

To be honest, I'm not sure whether the problem is with the plugin or AMQP library. I did look at the code and I don't think the problem is with AMQP plugin, but as I've said, I'm not sure. Please, advise.

Thank you,
Z.

**Comment by zstyblik** [comment]
Just to make sure, AMQP output is defined as:

```
[AMQPOutput]
message_matcher = "TRUE"
encoder = "json_payload"
url = "amqp://user:password@rmq.example.com/%2fvhost"
exchange = "amq.topic"
exchange_type = "topic"
exchange_durability = true
exchange_auto_delete = false
routing_key = "my.key"
content_type = "application/json"
use_framing = false

[AMQPOutput.retries]
max_retries = -1
```

**Comment by simonpasquier** [comment]
@zstyblik while trying to fix #1757 I get the same error message ("AMQP scheme must be either 'amqp://' or 'amqps://'"). AFAICT Heka hasn't changed the version of the AMQP library for a while so I suspect a code change in Heka v0.10 that breaks the init phase for the AMQP plugins when the pipeline runner wants to respawn a stopped plugin.

**Comment by zstyblik** [comment]
@simonpasquier it might be possible that plugin restart is somewhat wrong. I'm sorry, but I'm new to heka and I've only quickly read through AMQP plugin in heka. I've seen nothing that would suggest URL is being manipulated with. However, it's not impossible. I suspect IP address is somehow used instead of URL, but I currently have no means to verify that(yes, I have tried to avoid compilation of heka as much as possible :) ). I'm sort of glad to see I'm not alone who's having problems.

**Comment by pierce300** [comment]
+1

**Comment by simonpasquier** [comment]
I've dug further into this and confirmed that something is broken in the Heka framework wrt to the plugin restarts.

@rafrombrc I suspect that https://github.com/mozilla-services/heka/pull/1538 is the culprit (or maybe some subsequent commit). Affer this change, the Config() method of pluginMaker always returns the default ConfigStruct and not the ConfigStruct generated from the TOML. I hacked a dirty fix in my local repository [0] and I verified that the configuration is now correctly set when the plugin is respawned. I'd appreciate your feedback on whether my explanation makes sense and what would be the proper way to fix this. Thanks!

[0] https://github.com/simonpasquier/heka/commit/00143dce3f912b715eff53f31b872f2f042786c6

**Comment by samsieber** [comment]
+1 This bug is a showstopper for us where I work.

**Comment by rafrombrc** [comment]
@simonpasquier Yes, your assessment is absolutely correct. The PluginMaker was updated to no longer cache the config structs that have been populated by the TOML, but the plugin runners' restart code wasn't updated to account for this. Your fix is correct, `maker.PrepConfig()` should be used instead of `maker.Config()` in all of the restart loops. I'll update the code and will get the fix into the v0.10.0b2 release that I'm hoping to get out the door tomorrow.
